### DISCUSSION TITLE
Create transparent image with weather data for insertion into timelapse video

### DISCUSSION THREAD

**Discussion by JonFear** [Daveiano/weewx-wdc#151] [Ideas]
Hi Dave

Here's a good one...!

Take the OutTemp, OutHumidity, Mean Wind speed, Wind Direction, days rain total and create a transparent jpg to insert into a timelapse video using ffmpeg. update say every 5 minutes.

Any thoughts?

Jon

**Comment by Daveiano** [comment]
Hi Jon!

jpg files cant handle transparency, the generated image would need to be a png file.

I guess it shouldn't be too hard to generate an image like this since [pillow](https://pillow.readthedocs.io/en/stable/) is already required by weewx.

I will add this to the list and have a look but I will not super-prioritize this because it's a real edge case. But people will use it once included in the skin, I guess - I also like the idea.

**Comment by JonFear** [comment]
Dave

Understood. Looking at ffmpeg there is a way of inserting a text file. I wonder if I cam get the rss working in which case I suspect I could write up a script to format a text box. I need to think about it further. Still got to get ffmpeg create a file size which is manageable.. I will get it just need to play! So many option in ffmpeg it is almost impossible to know them all!

Jon

**Comment by Daveiano** [comment]
The RSS feed neets also to be provided by the skin or is there another way to generate one? The seasons skin does include a RSS template: https://github.com/weewx/weewx/blob/master/skins/Seasons/rss.xml.tmpl

**Comment by JonFear** [comment]
I was thinking of the Seasons version but obviously getting that to work at the same time as your skin maybe an issue. If an RSS feed can be "easily" added to your skin then that is the best way but I am very mindful of the work you already have on your list, mostly because of me...!

Jon

**Reply by Daveiano** [reply]
Oh no you can run multiple reports at a time, just give them a different `HTML_ROOT` in weewx.conf so they do not override each other.

A lot of people are using multiple skins at a time as far as I know.

**Comment by JonFear** [comment]
OK so some pruning in the standard report rss template then use xmlstarlet to gather the data...Simple script:

**weewx_rss.xml.tmpl**

```
<?xml version="1.0"?>
<rss version="2.0"
     xmlns:content="http://purl.org/rss/1.0/modules/content/"
     xmlns:geo="http://www.w3.org/2003/01/geo/wgs84_pos#" >
   <channel>
      <title>$station.location, $gettext("Weather Conditions")</title>
      <link>$station.station_url</link>
      <description>$gettext("Current conditions, and daily, monthly, and yearly summaries")</description>
      <language>"$lang"</language>
      <pubDate>$current.dateTime.format("%a, %d %b %Y %H:%M:%S %Z")</pubDate>

      <lastBuildDate>$current.dateTime.format("%a, %d %b %Y %H:%M:%S %Z")</lastBuildDate>
      <docs>http://blogs.law.harvard.edu/tech/rss</docs>
      <generator>weewx $station.version</generator>
      <ttl>$current.interval.string('')</ttl>

      <item>
         <title>$gettext(" Weather Conditions at") $current.dateTime; </title>
         <temp>$obs.label.outTemp: $current.outTemp; </temp>
         <baro>$obs.label.barometer: $current.barometer; </baro>
         <wind>$obs.label.wind: $current.windSpeed $gettext("from") $current.windDir; </wind>
         <rain>$gettext("Rain today"): $day.rain.sum; </rain>
      </item>

   </channel>
</rss>
```

Could be pruned further if needed just to get what is really needed.

Gather the data from the rss.xml file:

```
#!/bin/bash
xmlstarlet sel -t -v '//channel/item/title' /var/www/html/weewx/standard/RSS/weewx_rss.xml > rss-out.txt 
xmlstarlet sel -t -v '//channel/item/temp' /var/www/html/weewx/standard/RSS/weewx_rss.xml >> rss-out.txt 
xmlstarlet sel -t -v '//channel/item/baro' /var/www/html/weewx/standard/RSS/weewx_rss.xml >> rss-out.txt
xmlstarlet sel -t -v '//channel/item/wind' /var/www/html/weewx/standard/RSS/weewx_rss.xml >> rss-out.txt
xmlstarlet sel -t -v '//channel/item/rain' /var/www/html/weewx/standard/RSS/weewx_rss.xml >> rss-out.txt
```

Then make the variable file:

```
#!/bin/bash
rss_chain=$(cat /home/jon/rss-out.txt)
```

Then apply it to the ffmpeg output, creates an HD snapshot, adds to the dated timelapse directory and also places a copy for upload of the current image for upload to the site:

```
#!/bin/bash
datetoday=$(date +%Y%m%d)
mkdir -p /home/jon/nas/timelapse/nc-"$datetoday"
img_dir=/home/jon/nas/timelapse/nc-"$datetoday"
now=$(/bin/date '+%Y%m%d%H%M%S')
ffmpeg -rtsp_transport tcp -i rtsp://192.168.10.58:554/user=admin_password=nTBCS19C_channel=1_stream=0 -vf "drawtext=textfile=/home/jon/rss-out.txt: x=(w-text_w)/2:y=(h-th-20):fontsize=28:fontcolor=white" -frames:v 1 "$img_dir"/"${now}".jpg
cp "$img_dir"/"$now".jpg /var/www/html/weewx/snap/nc-snap-lrg.jpg
```

Simples really!

Oh and change the report generation time interval in weewx.conf to 1 minute. This updates the rss file every minute.

Jon

**Comment by Daveiano** [comment]
Hi Jon, 

I am happy that this is working for you!

Just a note: The RSS is not valid, for custom elements (like `temp`, `baro`) you need to namespace them (add a prefix) like this:

`<wdc:temp>$obs.label.outTemp: $current.outTemp; </wdc:temp>`

Also the opening `<RSS>` tag needs to be changed to: 

```
<rss version="2.0"
     xmlns:content="http://purl.org/rss/1.0/modules/content/"
     xmlns:geo="http://www.w3.org/2003/01/geo/wgs84_pos#" 
     xmlns:wdc="$station.station_url" 
>
```

Besides that, I am actually thinking about if there should be an RSS Feed included in the skin or not? People could probably use the seasons skin or if they really need to build a customized version as you did.

**Comment by JonFear** [comment]
Hi Dave

That is from the Standard report RSS template which I guess has not had the loving care given to it that it needs! As I am not using specifically for RSS purposes I will leave well alone. I have modded the original but cutting huge chunks out and renaming the bits I need naming but...

I tend to agree, no point in re-inventing the wheel when the functionality is pre-built-in. If there was a way to automate what I now do inside bash scripts then that would be an advantage but I guess would have seriously limited appeal, maybe even just one or two people! It does save having to use paid for software like IPTimelapse which also needs a windows machine to run on.

I think we can close out this discussion now.

Jon

**Comment by JonFear** [comment]
Discussion closed. Solutions found. Use Standard report skin for RSS feed. Use xmlstarlet for xml transformation and ffmpeg to insert text into video stream.
